### ISSUE TITLE
[FEAT] Support for zfs pools

### ISSUE THREAD

**Issue by alxsch** [AnalogJ/scrutiny#171]
Might be out of scope but maybe add zfs pool checking.
```zpool status```

**Comment by ariejan** [comment]
@AnalogJ what do you think about this? Personally I'm interested and _might_ be able to contribute some code for it.

**Comment by AnalogJ** [comment]
I'm definitely open to it, but I don't use ZFS personally so its a bit hard for me to develop for it, I'll need some help from the community. @ariejan I would definitely be open to looking at a PR.

Can someone paste the output of their `zpool status` command, just so I get a sense of the kind (and scope) of data involved?

**Comment by onedr0p** [comment]
[node_exporter](https://github.com/prometheus/node_exporter#enabled-by-default) already gives you zfs pool status and more. I don't think it's needed here.

**Comment by AnalogJ** [comment]
So I did some reading on `zpool` and I'm not quite sure how it'll fit in with Scrutiny's existing functionality. 
We'd have to design a new details page to support RAID (HW & SW) pool statuses, and visually separate out virtual vs physical devices in the Dashboard. 

I'm going to say that this is out-of-scope for now, since I don't have time to implement something like that (and I don't have hardware to test on). 

I'll keep this issue open as a place to discuss proposals and requirements for this feature.

**Comment by canDry** [comment]
> Can someone paste the output of their `zpool status` command, just so I get a sense of the kind (and scope) of data involved?

```
> zpool status
  pool: zfs-p1
 state: ONLINE
  scan: scrub repaired 0B in 0 days 04:27:37 with 0 errors on Sun May  8 04:51:38 2022
config:

	NAME        STATE     READ WRITE CKSUM
	zfs-p1      ONLINE       0     0     0
	  mirror-0  ONLINE       0     0     0
	    sda     ONLINE       0     0     0
	    sdb     ONLINE       0     0     0


```

**Comment by ariejan** [comment]
Here's one more:

```
treebeard:~# zpool status
  pool: mini-tank
 state: ONLINE
  scan: scrub repaired 0B in 02:12:17 with 0 errors on Sun May  8 02:36:19 2022
config:

	NAME                                        STATE     READ WRITE CKSUM
	mini-tank                                   ONLINE       0     0     0
	  ata-WDC_WD30EZRX-00D8PB0_WD-WMC4N0914861  ONLINE       0     0     0

errors: No known data errors

  pool: rpool
 state: ONLINE
  scan: scrub repaired 0B in 00:11:36 with 0 errors on Sun May  8 00:35:45 2022
config:

	NAME                                       STATE     READ WRITE CKSUM
	rpool                                      ONLINE       0     0     0
	  mirror-0                                 ONLINE       0     0     0
	    ata-CT250MX500SSD1_2208E60D104A-part3  ONLINE       0     0     0
	    ata-CT250MX500SSD1_1949E22CA8C7-part3  ONLINE       0     0     0

errors: No known data errors

  pool: tank
 state: ONLINE
  scan: scrub repaired 0B in 1 days 06:42:22 with 0 errors on Mon May  9 07:06:32 2022
config:

	NAME                                    STATE     READ WRITE CKSUM
	tank                                    ONLINE       0     0     0
	  raidz1-0                              ONLINE       0     0     0
	    wwn-0x50014ee2bb55b28f              ONLINE       0     0     0
	    wwn-0x50014ee265ff666a              ONLINE       0     0     0
	    wwn-0x50014ee210aaa174              ONLINE       0     0     0
	    wwn-0x50014ee265ff7131              ONLINE       0     0     0
	  raidz1-1                              ONLINE       0     0     0
	    wwn-0x5000cca257f07585              ONLINE       0     0     0
	    wwn-0x5000cca257f05822              ONLINE       0     0     0
	    wwn-0x5000cca257f01094              ONLINE       0     0     0
	    wwn-0x5000cca257f09d32              ONLINE       0     0     0
	  raidz1-2                              ONLINE       0     0     0
	    ata-WDC_WD140EMFZ-11A0WA0_9RK41S1L  ONLINE       0     0     0
	    ata-WDC_WD140EMFZ-11A0WA0_9RK0YKSL  ONLINE       0     0     0
	    wwn-0x5000cca264ec4096              ONLINE       0     0     0
	    wwn-0x5000cca264e7ce8a              ONLINE       0     0     0

errors: No known data errors

  pool: vms
 state: ONLINE
  scan: scrub repaired 0B in 01:29:23 with 0 errors on Sun May  8 01:53:35 2022
config:

	NAME                                  STATE     READ WRITE CKSUM
	vms                                   ONLINE       0     0     0
	  mirror-0                            ONLINE       0     0     0
	    ata-CT500MX500SSD1_2045E4C55CD0   ONLINE       0     0     0
	    ata-CT500MX500SSD1_2110E584D777   ONLINE       0     0     0
	  mirror-1                            ONLINE       0     0     0
	    ata-CT1000MX500SSD1_2205E605B1D4  ONLINE       0     0     0
	    ata-CT1000MX500SSD1_2205E605AF2A  ONLINE       0     0     0

errors: No known data errors
```

**Comment by almereyda** [comment]
While other commands that present a similar topology, like `zpool list -v` or `zpool iostat -v`, have an `-H` flag for tab-delimited, parsable output, `zpool status` can at times become a little messy and wordy for the sake of scraping:

```
$ zpool status
  pool: bpool
 state: ONLINE
  scan: scrub repaired 0B in 00:00:04 with 0 errors on Sun May  9 00:24:05 2021
config:

	NAME           STATE     READ WRITE CKSUM
	bpool          ONLINE       0     0     0
	  mirror-0     ONLINE       0     0     0
	    nvme2n1p3  ONLINE       0     0     0
	    nvme1n1p3  ONLINE       0     0     0

errors: No known data errors

  pool: ipool
 state: ONLINE
status: Some supported and requested features are not enabled on the pool.
	The pool can still be used, but some features are unavailable.
action: Enable all features using 'zpool upgrade'. Once this is done,
	the pool may no longer be accessible by software that does not support
	the features. See zpool-features(7) for details.
  scan: scrub repaired 0B in 2 days 03:07:51 with 0 errors on Tue May 11 03:31:53 2021
config:

	NAME                                            STATE     READ WRITE CKSUM
	ipool                                           ONLINE       0     0     0
	  mirror-0                                      ONLINE       0     0     0
	    ata-ST8000AS0002-1NA17Z_Z840BWA7            ONLINE       0     0     0
	    ata-ST8000AS0002-1NA17Z_Z840BN3P            ONLINE       0     0     0
	  mirror-1                                      ONLINE       0     0     0
	    ata-ST12000NM0008-2H3101_ZHZ1DAJG           ONLINE       0     0     0
	    ata-ST12000NM0008-2H3101_ZHZ55H00           ONLINE       0     0     0
	cache
	  nvme-KINGSTON_SA2000M81000G_50026B768473E536  ONLINE       0     0     0

errors: No known data errors

  pool: rpool
 state: ONLINE
status: Some supported and requested features are not enabled on the pool.
	The pool can still be used, but some features are unavailable.
action: Enable all features using 'zpool upgrade'. Once this is done,
	the pool may no longer be accessible by software that does not support
	the features. See zpool-features(7) for details.
  scan: scrub repaired 0B in 00:27:41 with 0 errors on Sun May  9 00:51:44 2021
config:

	NAME           STATE     READ WRITE CKSUM
	rpool          ONLINE       0     0     0
	  mirror-0     ONLINE       0     0     0
	    nvme2n1p4  ONLINE       0     0     0
	    nvme1n1p4  ONLINE       0     0     0

errors: No known data errors

```

Unfortunately the `-v` flag does not allow to filter the output in conjunction with `-o`, as with `-o name` for example, but it can be used to show `-H` parsable output and `-P` full device names:

```
$ zpool list -vHP    
bpool	1.88G	424M	1.46G	-	-	2%	22%	1.00x	ONLINE	-
	mirror	1.88G	424M	1.46G	-	-	2%	22.1%	-	ONLINE
	/dev/nvme2n1p3	-	-	-	-	-	-	-	-	ONLINE
	/dev/nvme1n1p3	-	-	-	-	-	-	-	-	ONLINE
ipool	18.2T	15.7T	2.49T	-	-	3%	86%	1.00x	ONLINE	-
	mirror	7.27T	7.19T	73.8G	-	-	8%	99.0%	-	ONLINE
	/dev/disk/by-id/ata-ST8000AS0002-1NA17Z_Z840BWA7-part1	-	-	-	-	-	-	-	-	ONLINE
	/dev/disk/by-id/ata-ST8000AS0002-1NA17Z_Z840BN3P-part1	-	-	-	-	-	-	-	-	ONLINE
	mirror	10.9T	8.49T	2.42T	-	-	0%	77.8%	-	ONLINE
	/dev/disk/by-id/ata-ST12000NM0008-2H3101_ZHZ1DAJG-part1	-	-	-	-	-	-	-	-	ONLINE
	/dev/disk/by-id/ata-ST12000NM0008-2H3101_ZHZ55H00-part1	-	-	-	-	-	-	-	-	ONLINE
cache                                                                     -      -      -        -         -      -      -      -  -
	/dev/disk/by-id/nvme-KINGSTON_SA2000M81000G_50026B768473E536-part1	931G	931G	103M	-	-	0%	100.0%	-	ONLINE
rpool	920G	284G	636G	-	-	31%	30%	1.00x	ONLINE	-
	mirror	920G	284G	636G	-	-	31%	30.8%	-	ONLINE
	/dev/nvme2n1p4	-	-	-	-	-	-	-	-	ONLINE
	/dev/nvme1n1p4	-	-	-	-	-	-	-	-	ONLINE

```

Similarly, a useful command to query for a system's pool's topologies, with gaining full device paths and displaying a syntax that keeps the association of cache devices distinctly with their pool, also seems to be:

```
$ zpool iostat -vPp   
                                                                                  capacity                         operations                         bandwidth           
pool                                                                            alloc             free             read            write             read            write
--------------------------------------------------------------------  ---------------  ---------------  ---------------  ---------------  ---------------  ---------------
bpool                                                                       444321792       1568944128                0                0               46              237
  mirror                                                                    444321792       1568944128                0                0               46              237
    /dev/nvme2n1p3                                                                  0                0                0                0               25              118
    /dev/nvme1n1p3                                                                  0                0                0                0               20              118
--------------------------------------------------------------------  ---------------  ---------------  ---------------  ---------------  ---------------  ---------------
ipool                                                                  17241778655232    2738409205760                1                4            43442            49020
  mirror                                                                7909430796288      79208374272                0                0            15608             1482
    /dev/disk/by-id/ata-ST8000AS0002-1NA17Z_Z840BWA7-part1                          0                0                0                0             7848              741
    /dev/disk/by-id/ata-ST8000AS0002-1NA17Z_Z840BN3P-part1                          0                0                0                0             7759              741
  mirror                                                                9332347858944    2659200831488                0                3            27834            47538
    /dev/disk/by-id/ata-ST12000NM0008-2H3101_ZHZ1DAJG-part1                         0                0                0                1            13630            23769
    /dev/disk/by-id/ata-ST12000NM0008-2H3101_ZHZ55H00-part1                         0                0                0                1            14204            23769
cache                                                                               -                -                -                -                -                -
  /dev/disk/by-id/nvme-KINGSTON_SA2000M81000G_50026B768473E536-part1    1000081873408        108110848                5                0            26788            21430
--------------------------------------------------------------------  ---------------  ---------------  ---------------  ---------------  ---------------  ---------------
rpool                                                                    304576651264     683265826816                4              108           129540          2415841
  mirror                                                                 304576651264     683265826816                4              108           129540          2415841
    /dev/nvme2n1p4                                                                  0                0                2               51            68274          1207920
    /dev/nvme1n1p4                                                                  0                0                1               56            61265          1207920
--------------------------------------------------------------------  ---------------  ---------------  ---------------  ---------------  ---------------  ---------------

```

With `-H` parsable, tab-delimited output, we'd lose the semantic white space, and the `cache` term altogether:

```
$ zpool iostat -vHPp
bpool	444321792	1568944128	0	0	46	237
mirror	444321792	1568944128	0	0	46	237
/dev/nvme2n1p3	0	0	0	0	25	118
/dev/nvme1n1p3	0	0	0	0	20	118
ipool	17241778667520	2738409193472	1	4	43373	49023
mirror	7909430796288	79208374272	0	0	15583	1482
/dev/disk/by-id/ata-ST8000AS0002-1NA17Z_Z840BWA7-part1	0	0	0	0	7836	741
/dev/disk/by-id/ata-ST8000AS0002-1NA17Z_Z840BN3P-part1	0	0	0	0	7746	741
mirror	9332347871232	2659200819200	0	3	27790	47540
/dev/disk/by-id/ata-ST12000NM0008-2H3101_ZHZ1DAJG-part1	0	0	0	1	13609	23770
/dev/disk/by-id/ata-ST12000NM0008-2H3101_ZHZ55H00-part1	0	0	0	1	14181	23770

/dev/disk/by-id/nvme-KINGSTON_SA2000M81000G_50026B768473E536-part1	1000081820160	108164096	5	0	26746	21396
rpool	304584048640	683258429440	4	108	129337	2416733
mirror	304584048640	683258429440	4	108	129337	2416733
/dev/nvme2n1p4	0	0	2	51	68166	1208366
/dev/nvme1n1p4	0	0	1	56	61170	1208366

```

**Comment by AnalogJ** [comment]
looks like JSON support was added to `zfs` and `zpool` ~2018

https://github.com/openzfs/zfs/pull/7281

**Comment by AnalogJ** [comment]
related https://github.com/AnalogJ/scrutiny/issues/329
related #415 

I'd be open to a PR adding a new collector for this kind of information, however it's not something I'll be able to take on myself.

**Comment by kaysond** [comment]
This issue is being closed automatically as a part of an effort to streamline and prioritize development effort. If it is still relevant, please reply to this comment and we will convert it to a discussion.

**Comment by girlpunk** [comment]
@kaysond still relevent
